pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+	// until the initial sync has been processed, and to avoid a hot
+	// loop, we'll wait this long between checks.
+	SyncProcessedPollPeriod = 100 * time.Millisecond
+
+	clientRetryCount    = 5
+	clientRetryInterval = 5 * time.Second
+)
+
+// IngressIPController is responsible for allocating ingress ip
+// addresses to Service objects of type LoadBalancer.
+type IngressIPController struct {
+	client kclient.ServicesNamespacer
+
+	// Services that need to be checked
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	cache      cache.Store
+	controller *framework.Controller
+
+	// Tracks ip allocation for the configured range
+	ipAllocator *ipallocator.Range
+	// Tracks ip -> service key to allow detection and correction of
+	// duplicate ip allocations.
+	allocationMap map[string]string
+	// Tracks allocations that will be performed after the initial sync
+	pendingAllocations []*kapi.Service
+
+	// Protects the transition between initial sync and regular processing
+	mu sync.RWMutex
+	// Tracks whether the initial sync has been processed
+	syncProcessed bool
+
+	// changeHandler does the work. It's factored out for unit testing
+	changeHandler func(change *serviceChange) error
+}
+
+type serviceChange struct {
+	key        string
+	oldService *kapi.Service
+}
+
+// NewIngressIPController creates a new IngressIPController.
+// TODO this should accept a shared informer
+func NewIngressIPController(client kclient.ServicesNamespacer, ipNet *net.IPNet, resyncInterval time.Duration) *IngressIPController {
+	ic := &IngressIPController{
+		client:     client,
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+
+	ic.cache, ic.controller = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return ic.client.Services(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return ic.client.Services(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&kapi.Service{},
+		resyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				service := obj.(*kapi.Service)
+				glog.V(4).Infof("Adding service %s/%s", service.Namespace, service.Name)
+				ic.enqueueChange(obj, nil)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				service := cur.(*kapi.Service)
+				glog.V(4).Infof("Updating service %s/%s", service.Namespace, service.Name)
+				ic.enqueueChange(cur, old)
+			},
+			DeleteFunc: func(obj interface{}) {
+				service := obj.(*kapi.Service)
+				glog.V(4).Infof("Deleting service %s/%s", service.Namespace, service.Name)
+				ic.enqueueChange(nil, obj)
+			},
+		},
+	)
+
+	ic.changeHandler = ic.processChange
+
+	ic.ipAllocator = ipallocator.NewAllocatorCIDRRange(ipNet, func(max int, rangeSpec string) allocator.Interface {
+		return allocator.NewContiguousAllocationMap(max, rangeSpec)
```

Comment:
I thought it was desirable to have relatively consistent allocation ordering to minimize the potential for conflicts between 2 competing controllers (i.e. when one is shutting down and the new one is coming online).  If that's not the case, is there a point to performing allocations from the initial sync in sorted order?